### PR TITLE
Allow multiple selections in report-list-filter UI

### PR DESCRIPTION
**Pull request by zarino**
To fix https://github.com/mysociety/fixmystreetforcouncils/issues/95

- [x] You can now select multiple statuses and categories in the UI
- [x] Update backend to work with multiple selections.
- [x] Fix test (probably will be by above)
- [x] Sort out vertical alignment across browsers
- [x] Sort out menu overflow in sidebar
   - https://github.com/mysociety/jquery-multi-select/pull/5

dropdowns

~~We should probably re-think the logic of the options here – for instance, should “all reports” be a fourth option in the status select box, or should the box just default to having its three options (“unfixed reports”, “closed reports”, “fixed reports”) ticked?~~

### CODE REVIEW

File: web/cobrands/fixmystreet/fixmystreet.js [hunk 1]
```diff
@@ -1223,11 +1223,16 @@ $(function() {
                     // location.href is something like foo.com/around?pc=abc-123,
                     // which we pass into fixmystreet.display.reports_list() as a fallback
                     // in case the list isn't already in the DOM.
+                    $('select[multiple] option').prop('selected', function() { return this.defaultSelected; }).trigger('change.multiselect');
```
Comment (zarino) [inline web/cobrands/fixmystreet/fixmystreet.js]: Should this be `$('#filter_categories option, #statuses option')` ?

File: web/cobrands/fixmystreet/fixmystreet.js [hunk 2]
```diff
@@ -1223,11 +1223,16 @@ $(function() {
...
                     fixmystreet.display.reports_list(location.href);
                 } else if ('reportId' in e.state) {
                     fixmystreet.display.report(e.state.reportPageUrl, e.state.reportId);
                 } else if ('newReportAtLonlat' in e.state) {
                     fixmystreet.display.begin_report(e.state.newReportAtLonlat, false);
+                } else if ('filter_change' in e.state) {
+                    $('#filter_categories').val(e.state.filter_change.filter_categories);
+                    $('#statuses').val(e.state.filter_change.statuses);
+                    $('select[multiple]').trigger('change.filters').trigger('change.multiselect');
```
Comment (zarino) [inline web/cobrands/fixmystreet/fixmystreet.js]: Should this be `$('#filter_categories, #statuses')` ?

Comment (dracos) [inline web/cobrands/fixmystreet/fixmystreet.js]: Sure (well, $('#filter_categories').add('#statuses') is about 10 times faster :) )

File: web/cobrands/fixmystreet/fixmystreet.js (hunk 1 above)
Comment (dracos) [inline web/cobrands/fixmystreet/fixmystreet.js]: Will go with `$('#filter_categories').add('#statuses').find('option');`

File: perllib/FixMyStreet/App/Controller/Around.pm [hunk 3]
```diff
@@ -258,13 +258,11 @@ sub check_and_stash_category : Private {
...
-
-    my $category = $c->get_param('filter_category');
     my %categories_mapped = map { $_ => 1 } @categories;
-    if ( defined $category && $categories_mapped{$category} ) {
-        $c->stash->{filter_category} = $category;
-    }
+
+    my $categories = [ $c->get_param_list('filter_category', 1) ];
+    my @valid_categories = grep { $_ && $categories_mapped{$_} } @$categories;
+    $c->stash->{filter_category} = \@valid_categories;
```
Comment (davea) [inline perllib/FixMyStreet/App/Controller/Around.pm]: I suppose renaming the request query param to `filter_categories` would run the risk of breaking bookmarks/links, but should the variable name used in the stash change to reflect its new expanded behaviour?

Comment (dracos) [inline perllib/FixMyStreet/App/Controller/Around.pm]: It's worse than that; filter_categories already exists, and is the list of categories to *show*.
Changing the request query param I'd be fine with, if not for that. I did try and change them all but got part way through and it all went horribly wrong so I rolled that bit back.

Comment (davea) [inline perllib/FixMyStreet/App/Controller/Around.pm]: Bah, nevermind then :)

**Comment by dracos** [comment]
I have updated the multi select plugin, so this presumably needs the FMS CSS updating and any new "contained within" ID giving, and then that should be it?

**Comment by zarino** [comment]
> this presumably needs the FMS CSS updating and any new "contained within" ID giving, and then that should be it?

Yes, I will do that right now.

**Pull request merged**